### DISCUSSION TITLE
Resizing a PNG image, working only with dictionary-based transforms

### DISCUSSION THREAD

**Discussion by hblanco2009** [Project-MONAI/MONAI#5347] [Q&A] [answered]
Hello guys. By following some of the examples online, I'm trying to load and resize a PNG image using Monai's transforms.
The following code, based on dictionary-based transforms, does the job fine, whereas the second one doesn't. 
I'm don't want to use these dictionary-based transforms, and I don't get where the problem is in the second example code.

# =========================================
# First example code -- IT WORKS!
# =========================================

from monai.transforms import (LoadImaged, EnsureChannelFirstd, Resized, EnsureTyped, Compose)

transform = Compose([LoadImaged(keys="image"),
                      EnsureChannelFirstd(keys="image"),
                      Resized(keys='image', spatial_size=[64, 64]),
                      EnsureTyped('image'),
                      ])

test_data = {'image': 'sample_image.png'}

result = transform(test_data)

print(f"image data shape:{result['image'].shape}")

# =========================================
# Similar example code -- IT DOESN'T WORK!
# =========================================

from monai.transforms import (LoadImage, EnsureChannelFirst, Resize, EnsureType, Compose)

transform = Compose([LoadImage(),
                      EnsureChannelFirst(),
                      Resize(spatial_size=[64, 64]),
                      EnsureType(),
                      ])

result = transform('sample_image.png')

print(f'image data shape:{result[0].shape}')

# =========================================
# Below a brief text of the console output
# =========================================

=== Transform input info -- EnsureChannelFirst ===
Data statistics:
Type: <class 'numpy.ndarray'> float32
Shape: (4280, 3520)
Value range: (7827.0, 63346.0)
Traceback (most recent call last):

 ... 

ValueError: meta_dict not available, EnsureChannelFirst is not in use.

**Comment by KumoLiu** [comment]
Hi @hblanco2009, may I know your MONAI version? I guess in your case, you can simply use `AddChannel` to add the channel. If you move to MONAI 1.0.0, you can use `EnsureChannelFirst` like this.
```
EnsureChannelFirst(channel_dim="no_channel")
```
Hope it can help you, thanks!

**Reply by hblanco2009** [reply]
Thank you @KumoLiu for your reply. Well this is the Monai's version installed in my computer: MONAI version: 0.8.0
Should I first uninstall the current version, just to move to Monai 1.0.0? Or there exist some kind of command, such as: Monai "update"?

I tried your suggestion: I replaced the "EnsureChannelFirst" transform by the "AsChannelFirst" transform, but the problem remains. Nonetheless, I debugged deeper the example code and find out what the problem is. I don't know if this is an issue or not, a bug or not. But here you are what I find out.
- When the "LoadImage" transform is applied on a PNG image the outcome is a tuple composed by 1) the image matrix and 2) a dictionary with metadata describing the image features (e.g., dimension, mode, format, etc.). Afterwards, this tuple is passed to the AddChannel transform. It works fine when applied on the image matrix, but fails when trying to add a channel to dictionary associated to the image matrix, that is the second tuple component.

Any suggestion on how to solve this? how do I pass the aforementioned tuple's first component to the AddChannel transform? How do I get rid of the dictionary?

**Reply by KumoLiu** [reply]
Hi @hblanco2009, you can update your MONAI version to 1.0.0 to solve this problem by using:
```
pip install monai==1.0.0
```
Thanks!

**Comment by hblanco2009** [comment, marked as answer]
Dear @KumoLiu, I successfully reinstalled Monai as you suggested. Afterwards, I just substituted the transform "AddChannel" (which is deprecated in this 1.0.0 version) by the suggested transform: EnsureChannelFirst(channel_dim="no_channel"). It is happening exactly the same as with the previous version, despite of the new Monai version. That is, the LoadImage() transform returns a tuple when reading a PNG image. The first component is now a MetaTensor (which contains the pixel matrix), whereas the second component is a dictionary. The EnsureChannelFirst transform is applied to both components, but fails on the dictionary. In the image below, you can see that when calling the EnsureChannelFirst over the object "img" (i.e., img = MetaTensor(img)), which is actually a dictionary, as you can see in the screenshot below, it fails and returns this messge: *** RuntimeError: Could not infer dtype of dict

![issue_with_PNG_images](https://user-images.githubusercontent.com/61469912/196429824-34443706-9ead-4596-8df5-cf32471e4914.png)

**Reply by KumoLiu** [reply]
Hi @hblanco2009, you can load the image with `image_only=True` like this:
```
img = LoadImage(image_only=True)(img)
img = EnsureChannelFirst(channel_dim="no_channel")(img)
```
Hope it can help you, thanks!

**Reply by hblanco2009** [reply]
Yes!!!!, .. now it works perfectly. Thank you so much!

**Reply by vikashg** [reply]
I have an additional question regarding the behavior of `EnsureChannelFirst`. My understanding is if there is an image of shape (nx, ny, 3), using `EnsureChannelFirst(channel_dim=2)` should give me an image of shape (3, nx, ny). But I get an image of shape (nx, ny). 
Any insights @KumoLiu

**Reply by KumoLiu** [reply]
Hi @vikashg, your understanding is exactly right. It will return an image of the shape (3, nx, ny). You can try the example code below.
```
from monai.transforms import EnsureChannelFirst
import torch

data = torch.zeros([64,64,3])
data = EnsureChannelFirst(channel_dim=2)(data)
print(data.shape) --> (3, 64, 64)
```
You can also refer to the docstring of `channel_dim`.
https://github.com/Project-MONAI/MONAI/blob/5b3b74143f9c8b985974e583b0a6f4029eef50e1/monai/transforms/utility/array.py#L212-L215
Hope it can help you, thanks!

**Reply by vikashg** [reply]
But, it doesnot do that for JPG image. Can you please try running on this [image](https://github.com/Project-MONAI/model-zoo/blob/vikash/149-breastdensity/models/breast_density_classification/sample_data/A/sample_A1.jpg)

```
transform_list = Compose([LoadImage(image_only=True), EnsureChannelFirst(channel_dim=2)])
img = transfrom_list(image_name)
print(img.shape)
```

**Reply by KumoLiu** [reply]
Hi @vikashg, it also works.
```
from monai.transforms import LoadImage, EnsureChannelFirst

image_name = '/workspace/sample_A1.jpeg'
image = LoadImage(image_only=True)(image_name)
print(image.shape) --> (1996, 2457, 3)
image = EnsureChannelFirst(channel_dim=2)(image)
print(image.shape) --> (3, 1996, 2457)
```

**Reply by vikashg** [reply]
@KumoLiu Sorry my bad. I had a bug in my code. However, it arises from my understanding of metatensor and I am trying to transition my code to make sure that it works. 
My understanding is that in the above code the resulting `image` is a MetaTensor. If `image` is a MetaTensor then it is a tuple where the first item is an array and the second item is the meta information. Is this a correct statement ?

If the first term is an array, i was doing `print(image[0].shape)` and I was just getting the two dimensions ? I feel I am missing something here.

**Reply by vikashg** [reply]
Or is it write to say that only `LoadImage` gives metatensor when I do not use `image_only=True` and `EnsureChannelFirst` cannot take metatensor as input ?

**Reply by KumoLiu** [reply]
Hi @vikashg, `image` is a MetaTensor is right, but it is not tuple, it's type is "<class 'monai.data.meta_tensor.MetaTensor'>".  MetaTensor is a subclass of `torch.Tensor`, most of the PyTorch APIs are supported. You can get meta information by using `image.meta` and for more use guide you can refer to [here](https://colab.research.google.com/drive/1T4iAys-cC2qL80oJkIbAXAPlWNPwp4H7).

When you use `LoadImage` with `image_only=True`, it will only return a MetaTonsor, and when it is with `image_only=False`, it will return a tuple, first is a MetaTensor, and second is `image.meta` which is meta information from MetaTensor. 

`EnsureChannelFirst ` can take MetaTensor as input.
```
image = LoadImage(image_only=True)(image_name) --> return a MetaTensor
image = EnsureChannelFirst(channel_dim=2)(image)

or

image = LoadImage(image_only=False)(image_name) --> return (MetaTensor, meta)
image = EnsureChannelFirst(channel_dim=2)(image[0])
```

Thanks!

**Reply by vikashg** [reply]
Hi @KumoLiu, Ok thanks for the clarification. I think the confusion arises from this piece of code. Lets say we have a `list_of_image_names`.  and a transform list defined as 
```
transform_list = Compose([LoadImage(image_only=True), EnsureChannelFirst(channel_dim=2)])
image_list = transform_list(list_of_image_names)
``` 
Now, if I print the `image_list`, it prints the image tensor as well as the metadata as a list which I believe is a list of metatensor. Now, if I go by that logic `image[0]` should give me the first metatensor (the array + the metadata). But, if I do `print(image[0])`, it just prints the image array. If I do `print(image[0].shape)`, it prints the correct image shape. 
So, if it seems like the metatensor behaves like a torch tensor sometimes but sometimes it acts as a metatensor. This distinction is not very clear. Is there an example for a simple training example, that I can refer. I know there are tutorials and I am not sure if they have been updated for handling metatensors. 

Thanks

**Reply by KumoLiu** [reply]
Hi @vikashg, all the tutorials are updated for MetaTensor, feel free to refer to any of them. Thanks!
